plugins/org.eclipse.n4js.flowgraphs/src/org/eclipse/n4js/flowgraphs/analysers/NeverUsedPredicateWalker.java

```diff
@@ -0,0 +1,110 @@
+/**
+ * Copyright (c) 2017 NumberFour AG.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *   NumberFour AG - Initial API and implementation
+ */
+package org.eclipse.n4js.flowgraphs.analysers;
+
+import org.eclipse.n4js.flowgraphs.FlowEdge;
+import org.eclipse.n4js.flowgraphs.analyses.GraphVisitor;
+import org.eclipse.n4js.n4JS.ControlFlowElement;
+import org.eclipse.n4js.n4JS.IdentifierRef;
+import org.eclipse.n4js.n4JS.VariableDeclaration;
+
+/**
+ * Checks if a local variable is never used.
+ */
+public class NeverUsedPredicateWalker extends GraphVisitor {
+	// TODO GH-235
+
+	NeverUsedPredicateWalker() {
+		super(Direction.Forward);
+	}
+
+	@Override
+	protected void initAll() {
+		// nothing to do
+	}
+
+	@Override
+	protected void init(Direction curDirection, ControlFlowElement curContainer) {
+		// nothing to do
+	}
+
+	@Override
+	protected void terminate(Direction curDirection, ControlFlowElement curContainer) {
+		// nothing to do
+	}
+
+	@Override
+	protected void terminateAll() {
+		// nothing to do
+	}
+
+	@Override
+	protected void visit(ControlFlowElement cfe) {
+		if (cfe instanceof IdentifierRef && cfe.eContainer() instanceof VariableDeclaration) {
+			super.requestActivation(new NeverUsedActivatedPathPredicate((IdentifierRef) cfe));
+		}
+	}
+
+	@Override
+	protected void visit(ControlFlowElement start, ControlFlowElement end, FlowEdge edge) {
+		// nothing to do
+	}
+
+	private class NeverUsedActivatedPathPredicate extends PathExplorer {
+		@SuppressWarnings("unused")
+		final IdentifierRef idRef;
+
+		public NeverUsedActivatedPathPredicate(IdentifierRef idRef) {
+			super(PredicateType.ForOnePath);
+			this.idRef = idRef;
+		}
+
+		@Override
+		protected NeverUsedActivePath firstPath() {
+			return new NeverUsedActivePath();
+		}
+
+		private class NeverUsedActivePath extends PathWalker {
```

Comment:
nested class in the nested class?

Reply:
_... that's double the fun!_

In my opinion, this could be actually helpful when implementing some analyses.

Reply:
I extracted the PathExplorer(-Internal) classes